test/servers/rpc-router.test.ts

```diff
@@ -0,0 +1,102 @@
+import 'mocha';
+import * as express from 'express';
+import * as sinon from 'sinon';
+import {RPCRouter, RPCServer} from '../../src/servers';
+import {expect} from '@loopback/testlab';
+
+describe('rpc-router', () => {
+  let router: RPCRouter;
+  let request: express.Request;
+  let response: express.Response;
+  // tslint:disable-next-line:no-any
+  let controller: any;
+
+  beforeEach(setupFakes);
+  it('routes to existing controller and method', async () => {
+    await router.routeRequest(request, response);
+    const stub = response.send as sinon.SinonStub;
+    expect(stub.called);
+    const result = stub.firstCall.args[0];
+    expect(result).to.match(/Hello, World!/);
+  });
+
+  it('returns error if controller does not exist', async () => {
+    request = getRequest({
+      body: {
+        controller: 'NotAController',
+      },
+    });
+    controller.rejects(new Error('Does not exist!'));
+    await router.routeRequest(request, response);
+    const stub = response.send as sinon.SinonStub;
+    expect(stub.called);
+    expect(response.statusCode).to.equal(400);
+    const result = stub.firstCall.args[0];
+    expect(result).to.match(/Does not exist!/);
+  });
+
+  it('returns error if method does not exist', async () => {
+    request = getRequest({
+      body: {
+        controller: 'FakeController',
+        method: 'notReal',
+        input: {
+          name: 'World',
+        },
+      },
+    });
+    await router.routeRequest(request, response);
+    const stub = response.send as sinon.SinonStub;
+    expect(stub.called);
+    expect(response.statusCode).to.equal(400);
+    const result = stub.firstCall.args[0];
+    expect(result).to.match(/No method was found on controller/);
+  });
+
+  function getRouter() {
+    const server = sinon.createStubInstance(RPCServer);
+    return new RPCRouter(server);
+  }
+
+  function getController(rtr: RPCRouter) {
+    const stub = sinon.stub(rtr, 'getController');
+    stub.resolves(new FakeController());
+    return stub;
+  }
+
+  function getRequest(req?: Partial<express.Request>) {
+    const reqt = Object.assign(
+      <express.Request>{
```

Comment:
Uh oh, this is hacky and may easily break when express starts to require on another request property that's not mocked out here, or calls any of the ReadableStream APIs. Have you considered using [Shot](https://www.npmjs.com/package/shot) instead? We are shipping it as a part of our testlab, see [packages/testlab/src/shot.ts](https://github.com/strongloop/loopback-next/blob/df1c8799f681af30680a390f5d8081290c9047e7/packages/testlab/src/shot.ts)

Reply:
@bajtos Would you be able to provide an example of how to use our testlab exports of Shot? I've noticed that we're not actually using it the way Shot's Github README advertises, and I'm not certain how I would use it to mock the handling of express requests and responses (given that it's for `http.ServerRequest` and `http.ServerResponse`).

I understand that the hard cast here could break if something leverages a function or property that I haven't faked, but I've gone out of my way to make sure that nothing calls functions on the request or response object in the rpc-router code above and beyond what's faked here.

Reply:
Technically speaking, no instance of an express server ever begins listening on an external port; these tests are essentially integration tests.

Reply:
> Would you be able to provide an example of how to use our testlab exports of Shot?

Have you tried to search for the places in loopback-next that are using it? There are quire few test files leveraging `mockResponse`, see https://github.com/strongloop/loopback-next/search?utf8=%E2%9C%93&q=shot&type=

Looks like we need to improve the docs for this part of testlab! I created an issue - see https://github.com/strongloop/loopback-next/issues/760